packages/protocol/contracts/stability/Broker.sol

```diff
@@ -0,0 +1,229 @@
+   * @param _reserve The Mento reserve address
+   */
+  function setReserve(address _reserve) public onlyOwner {
+    require(_reserve != address(0), "Reserve address must be set");
+    emit ReserveSet(_reserve, address(reserve));
+    reserve = IReserve(_reserve);
+  }
+
+  /**
+   * @notice Calculate amountIn of tokenIn for a given amountIn of tokenIn
+   * @param exchangeManager the address of the exchange manager for the pair
+   * @param exchangeId The id of the exchange to use
+   * @param tokenIn The token to be sold
+   * @param tokenOut The token to be bought
+   * @param amountOut The amount of tokenOut to be bought
+   * @return amountIn The amount of tokenIn to be sold
+   */
+  function getAmountIn(
+    address exchangeManager,
+    bytes32 exchangeId,
+    address tokenIn,
+    address tokenOut,
+    uint256 amountOut
+  ) external returns (uint256 amountIn) {
+    require(checkIsExchangeManager(exchangeManager), "ExchangeManager does not exist");
+    return
+      amountOut = IExchangeManager(exchangeManager).getAmountIn(
+        exchangeId,
+        tokenIn,
+        tokenOut,
+        amountOut
+      );
+  }
+
+  /**
+   * @notice Calculate amountOut of tokenOut for a given amountIn of tokenIn
+   * @param exchangeManager the address of the exchange manager for the pair
+   * @param exchangeId The id of the exchange to use
+   * @param tokenIn The token to be sold
+   * @param tokenOut The token to be bought
+   * @param amountIn The amount of tokenIn to be sold
+   * @return amountOut The amount of tokenOut to be bought
+   */
+  function getAmountOut(
+    address exchangeManager,
+    bytes32 exchangeId,
+    address tokenIn,
+    address tokenOut,
+    uint256 amountIn
+  ) external returns (uint256 amountOut) {
+    require(checkIsExchangeManager(exchangeManager), "ExchangeManager does not exist");
+    return
+      amountOut = IExchangeManager(exchangeManager).getAmountOut(
+        exchangeId,
+        tokenIn,
+        tokenOut,
+        amountIn
+      );
+  }
+
+  /**
+   * @notice Execute a token swap with fixed amountIn
+   * @param exchangeManager the address of the exchange manager for the pair
+   * @param exchangeId The id of the exchange to use
+   * @param tokenIn The token to be sold
+   * @param tokenOut The token to be bought
+   * @param amountIn The amount of tokenIn to be sold
+   * @param amountOutMin Minimum amountOut to be received - controls slippage
+   * @return amountOut The amount of tokenOut to be bought
+   */
+  function swapIn(
+    address exchangeManager,
+    bytes32 exchangeId,
+    address tokenIn,
+    address tokenOut,
+    uint256 amountIn,
+    uint256 amountOutMin
+  ) external returns (uint256 amountOut) {
+    require(checkIsExchangeManager(exchangeManager), "ExchangeManager does not exist");
+    return
+      amountOut = IExchangeManager(exchangeManager).swapIn(
+        exchangeId,
+        tokenIn,
+        tokenOut,
+        amountIn,
+        amountOutMin
+      );
```

Comment:
There's one more thing here. I know this isn't currently reflected in the interface but we decided to do slippage control on the Broker rather than the exchange manager which means that I'd need you to do two things on your branch:

1. Edit `IExchangeManager.sol` and remove the `amountOutMin` and `amountInMax` from the `swapIn` and `swapOut` functions.
2. Here add an additional require for `amountOut >= amountOutMin` and revert with "amountOutMin not met" and similarly in `swapOut` have `amountIn <= amountInMax` and revert with "amountInMax exceeded"

Makes sense?

Reply:
yes that was done in the order version